Demos/Shopping/ShoppingUITests/TestPages/MainPage.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  MainPage.swift
+//  ShoppingUITests
+//
+//  Created by Albert Bori on 1/6/23.
+//
+
+import XCTest
+
+/// The main root app view which also supplies the tab bar and the default view
+struct MainPage: TabbedPage {
+    let app: XCUIApplication
+    
+    @discardableResult
+    func defaultTab(file: StaticString = #file, line: UInt = #line) -> ProductsTabPage {
+        .init(app: app, file: file, line: line)
+    }
+}
+
+/// The default tab that shows products
+struct ProductsTabPage: TabbedPage, CartButtonProviding {
+    let app: XCUIApplication
+    
+    init(app: XCUIApplication, file: StaticString = #file, line: UInt = #line) {
+        self.app = app
+        XCTAssertTrue(app.scrollViews.element(boundBy: 0).waitForExistence(), "Can't find Products scroll view", file: file, line: line)
+    }
+    
+    @discardableResult
+    func assertProductsPageIsVisible(file: StaticString = #file, line: UInt = #line) -> Self {
+        assert(app.navigationBars["Products"].exists, message: "Can't find 'Products' nav bar", file: file, line: line)
+    }
+    
+    @discardableResult
+    func tapProductCell(for product: TestProduct, file: StaticString = #file, line: UInt = #line) -> ProductDetailPage {
+        XCTAssertTrue(app.buttons[product.name].exists, "Can't find product cell '\(product.name)' button", file: file, line: line)
+        app.buttons[product.name].tap()
+        return .init(app: app, previousView: self, product: product, file: file, line: line)
+    }
+}
+
+/// The second tab view that shows account options
+struct AccountTabPage: TabbedPage, CartButtonProviding {
+    let app: XCUIApplication
+    
+    init(app: XCUIApplication, file: StaticString = #file, line: UInt = #line) {
+        self.app = app
+        XCTAssertTrue(app.navigationBars["Account"].waitForExistence(), "Can't find Account nav bar item", file: file, line: line)
+    }
+    
+    @discardableResult
+    func assertAccountPageIsVisible(file: StaticString = #file, line: UInt = #line) -> Self {
+        assert(app.navigationBars["Account"].exists, message: "Can't find 'Account' nav bar", file: file, line: line)
+    }
+    
+    @discardableResult
+    func tapFavorites(file: StaticString = #file, line: UInt = #line) -> FavoritesPage {
+        XCTAssertTrue(app.buttons["Favorites"].exists, "Can't find Favorites button", file: file, line: line)
+        app.buttons["Favorites"].tap()
+        return .init(app: app, previousView: self, file: file, line: line)
+    }
+    
+    @discardableResult
+    func tapSettings(file: StaticString = #file, line: UInt = #line) -> SettingsPage {
+        XCTAssertTrue(app.buttons["Settings"].exists, "Can't find Settings button", file: file, line: line)
```

Comment:
maybe find and tap could be combined to one call everywhere? would be nicely reuseable..

Reply:
I agree with this if asserting the element exists or isHittable would be better

Reply:
For a larger scale project, I would agree as well. However, considering that some elements need to be waited for, or have other pre-conditions prior to being tapped, the function code itself would offset the savings for the volume of scenarios in this test suite.

If it's alright with you, I think I'll defer on this suggestion. Other changes I made already significantly improve the readability.